File: testbed_php/test/associations/ManyToOptionalNTest.php
<?php

class ManyToOptionalNTest extends UnitTestCase
{


  public function test_CreateStudentWithoutMentor()
  {
    $s = new StudentAA(99);
    $this->assertEqual(0,$s->numberOfMentors());
  }
  

  public function test_CreateMentorWithoutStudent()
  {
    $m = new MentorAA("blah");
    $this->assertEqual(0,$m->numberOfStudents());
  }


  public function test_setTooMany()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);
    $m = new MentorAA("blah");
    $this->assertEqual(false,$m->setStudents(array($s,$s2,$s3))); 
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual(0,$s2->numberOfMentors());
    $this->assertEqual(0,$s3->numberOfMentors());
  }
  
  

  public function test_setStudentsAtMax()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);

    $m = new MentorAA("blah");
    $m->setStudents(array($s,$s2));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
  }
  

  public function test_setStudentsWatchOutForDuplicates()
  {
    $s = new StudentAA(99);
    $m = new MentorAA("blah");
    $this->assertEqual(false,$m->setStudents(array($s,$s)));
  }
  

  public function test_setStudentsDoNotEraseExisting()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);
    $s3 = new StudentAA(97);
    
    $m = new MentorAA("blah");
    $m->setStudents(array($s,$s2));

    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());    

    $m2 = new MentorAA("blah");
    $m2->setStudents(array($s,$s3));
    
    $this->assertEqual(2, $m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    
    $this->assertEqual(2, $m2->numberOfStudents());
    $this->assertEqual($m2,$s->getMentor(1));
    $this->assertEqual($m2,$s3->getMentor(0));

  }   
  

  public function test_setStudents_aboveNKeepsExistingIntact()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);
    $s4 = new StudentAA(96);
    $s5 = new StudentAA(95);

    $m = new MentorAA("blah");
    
    $this->assertEqual(true,$m->setStudents(array($s,$s2)));
    $this->assertEqual(false, $m->setStudents($s2,$s3,$s4,$s5,$s));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual(0,$s3->numberOfMentors());
    $this->assertEqual(0,$s4->numberOfMentors());
    $this->assertEqual(0,$s5->numberOfMentors());

  }


  public function test_setStudents_doNotAllowDuplicatesNM()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    

    $m = new MentorAA("blah");
    $this->assertEqual(false, $m->setStudents(array($s,$s2,$s)));
  }


  public function test_setStudents_atNOverwritesExisting()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);

    $m = new MentorAA("blah");
    
    $this->assertEqual(true, $m->setStudents(array($s,$s2)));
    $this->assertEqual(true, $m->setStudents(array($s2,$s3)));
    $this->assertEqual(2,$m->numberOfStudents());

    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual($m,$s2->getMentor(0));
    $this->assertEqual($m,$s3->getMentor(0));
   
  }
  

  public function test_addMentorTooMany()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);

    $m = new MentorAA("blah");
    $m->setStudents(array($s,$s2));
    
    $this->assertEqual(false,$s3->addMentor($m));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(0,$s3->numberOfMentors());
    
  }


  public function test_addMentorTOkay()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    

    $m = new MentorAA("blah");
    $m->setStudents(array($s));
    
    $this->assertEqual(true,$s2->addMentor($m));
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(1,$s2->numberOfMentors());
    
  }
  
  

  public function test_addStudent()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);
    $s3 = new StudentAA(97); 

    $m = new MentorAA("blah");
    $this->assertEqual(true,$m->setStudents(array($s)));

    $this->assertEqual(true,$m->addStudent($s2));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
   
    $this->assertEqual(false,$m->addStudent($s3));
    $this->assertEqual(2,$m->numberOfStudents());
  
  }  


  public function test_addStudentHasExistingMentor()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);

    $m = new MentorAA("blah");
    $m2 = new MentorAA("blah2");

    $m->setStudents(array($s,$s2));
    $m2->setStudents(array($s3));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(1,$m2->numberOfStudents());

    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual($m,$s2->getMentor(0));
    
    $this->assertEqual($m2,$s3->getMentor(0));
    
    $this->assertEqual(true,$m2->addStudent($s2));
    
    $this->assertEqual(2,$m->numberOfStudents());
    $this->assertEqual(2,$m2->numberOfStudents());
    
    $this->assertEqual(2,$s2->numberOfMentors());    
    $this->assertEqual($m,$s2->getMentor(0));    
    $this->assertEqual($m2,$s2->getMentor(1));

    $this->assertEqual(false,$m->addStudent($s3));
    $this->assertEqual(2,$s2->numberOfMentors());    
  }  


  public function test_removeStudent()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);    
    $s3 = new StudentAA(97);

    $m = new MentorAA("blah");
    $m->setStudents(array($s,$s2));
    
    $this->assertEqual(false,$m->removeStudent($s3));
    $this->assertEqual(true,$m->removeStudent($s2));
    
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($m,$s->getMentor(0));
    $this->assertEqual(0,$s2->numberOfMentors());
    $this->assertEqual(0,$s3->numberOfMentors());

    $this->assertEqual(true,$m->removeStudent($s));
    $this->assertEqual(0,$m->numberOfStudents());
  }  

  public function test_setStudents_empty()
  {
    $s = new StudentAA(99);
    $s2 = new StudentAA(98);
    $m = new MentorAA("blah");
    $m->setStudents(array($s,$s2));
    $this->assertEqual(true, $m->setStudents(array()));
    $this->assertEqual(0,$m->numberOfStudents());
    $this->assertEqual(0,$s->numberOfMentors());
    $this->assertEqual(0,$s2->numberOfMentors());
  }

  

  public function test_deleteStudentLeavesMentorAlone()
  {
    $s1 = new StudentAA(99);
    $s2 = new StudentAA(98);

    $m = new MentorAA("blah");
    $m->setStudents(array($s1,$s2));
    $studentP = new ProgramAA();
    $mentorP = new ProgramAA();
    
    $s2->setProgram($studentP);
    $m->setProgram($mentorP);

    $s1->delete();
    
    $this->assertEqual(1,$m->numberOfStudents());
    $this->assertEqual($mentorP,$m->getProgram());
    $this->assertEqual($m,$mentorP->getMentor());
    $this->assertEqual($studentP,$s2->getProgram());
    $this->assertEqual(0,$s1->numberOfMentors());
  } 
  

  public function test_studentBounds()
  {
    $this->assertEqual(2,MentorAA::maximumNumberOfStudents());
  }

  
  
}
?>

File: testbed_ruby/test/associations/many_to_optional_n_test.rb
require 'test/unit'
require './src-gen-umple/student_a_a'
require './src-gen-umple/mentor_a_a'
require './src-gen-umple/program_a_a'

module CruiseTest
class ManyToOptionalNTest < Test::Unit::TestCase

  def test_CreateStudentWithoutMentor
    s = StudentAA.new(99)
    assert_equal(0,s.number_of_mentors)
  end

  def test_CreateMentorWithoutStudent
  
    m = MentorAA.new("blah")
    assert_equal(0,m.number_of_students)
  end

  def test_setTooMany
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)
    m = MentorAA.new("blah")
    assert_equal(false,m.set_students([s,s2,s3])) 
    assert_equal(0,m.number_of_students)
    assert_equal(0,s.number_of_mentors)
    assert_equal(0,s2.number_of_mentors)
    assert_equal(0,s3.number_of_mentors)
  end


  def test_set_studentsAtMax
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)

    m = MentorAA.new("blah")
    m.set_students([s,s2])

    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
  end

  def test_set_studentsWatchOutForDuplicates
  
    s = StudentAA.new(99)
    m = MentorAA.new("blah")
    assert_equal(false,m.set_students([s,s]))
  end

  def test_set_studentsDoNotEraseExisting
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)
    s3 = StudentAA.new(97)

    m = MentorAA.new("blah")
    m.set_students([s,s2])

    assert_equal(2, m.number_of_students)
    assert_equal(m, s.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))
    assert_equal(0, s3.number_of_mentors)    

    m2 = MentorAA.new("blah2")
    m2.set_students([s,s3])

    assert_equal(2, m.number_of_students)
    assert_equal(m, s.get_mentor(0))
    assert_equal(m, s2.get_mentor(0))

    assert_equal(2, m2.number_of_students)
    assert_equal(m2, s.get_mentor(1))
    assert_equal(m2, s3.get_mentor(0))

  end   

  def test_set_students_aboveNKeepsExistingIntact
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)
    s4 = StudentAA.new(96)
    s5 = StudentAA.new(95)

    m = MentorAA.new("blah")

    assert_equal(true,m.set_students([s,s2]))
    assert_equal(false, m.set_students([s2, s3, s4, s5, s]))
    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))
    assert_equal(0,s3.number_of_mentors)
    assert_equal(0,s4.number_of_mentors)
    assert_equal(0,s5.number_of_mentors)

  end

  def test_set_students_doNotAllowDuplicatesNM
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    

    m = MentorAA.new("blah")
    assert_equal(false, m.set_students([s, s2, s]))
  end

  def test_set_students_atNOverwritesExisting
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)

    m = MentorAA.new("blah")

    assert_equal(true, m.set_students([s, s2]))
    assert_equal(true, m.set_students([s2, s3]))
    assert_equal(2,m.number_of_students)

    assert_equal(0,s.number_of_mentors)
    assert_equal(m,s2.get_mentor(0))
    assert_equal(m,s3.get_mentor(0))

  end

  def test_add_mentorTooMany
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)

    m = MentorAA.new("blah")
    m.set_students([s,s2])

    assert_equal(false,s3.add_mentor(m))
    assert_equal(2,m.number_of_students)
    assert_equal(0,s3.number_of_mentors)

  end

  def test_add_mentorTOkay
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    

    m = MentorAA.new("blah")
    m.set_students([s])

    assert_equal(true,s2.add_mentor(m))
    assert_equal(2,m.number_of_students)
    assert_equal(1,s2.number_of_mentors)

  end


  def test_add_student
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)
    s3 = StudentAA.new(97) 

    m = MentorAA.new("blah")
    assert_equal(true,m.set_students([s]))

    assert_equal(1,m.number_of_students)

    assert_equal(true,m.add_student(s2))

    assert_equal(2,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))

    assert_equal(false,m.add_student(s3))
    assert_equal(2,m.number_of_students)

  end  

  def test_add_studentHasExistingMentor
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)

    m = MentorAA.new("blah")
    m2 = MentorAA.new("blah2")

    m.set_students([s,s2])
    m2.set_students([s3])

    assert_equal(2,m.number_of_students)
    assert_equal(1,m2.number_of_students)

    assert_equal(m,s.get_mentor(0))
    assert_equal(m,s2.get_mentor(0))

    assert_equal(m2,s3.get_mentor(0))

    assert_equal(true,m2.add_student(s2))

    assert_equal(2,m.number_of_students)
    assert_equal(2,m2.number_of_students)

    assert_equal(2,s2.number_of_mentors)    
    assert_equal(m,s2.get_mentor(0))    
    assert_equal(m2,s2.get_mentor(1))

    assert_equal(false,m.add_student(s3))
  end  

  def test_remove_student
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)    
    s3 = StudentAA.new(97)

    m = MentorAA.new("blah")
    m.set_students([s,s2])

    assert_equal(false,m.remove_student(s3))
    assert_equal(true,m.remove_student(s2))

    assert_equal(1,m.number_of_students)
    assert_equal(m,s.get_mentor(0))
    assert_equal(0,s2.number_of_mentors)
    assert_equal(0,s3.number_of_mentors)

    assert_equal(true,m.remove_student(s))
    assert_equal(0,m.number_of_students)
  end  




  def test_set_students_empty
  
    s = StudentAA.new(99)
    s2 = StudentAA.new(98)
    m = MentorAA.new("blah")
    m.set_students([s,s2])
    assert_equal(true, m.set_students([]))
    assert_equal(0,m.number_of_students)
    assert_equal(0,s.number_of_mentors)
    assert_equal(0,s2.number_of_mentors)
  end


  def test_deleteStudentLeavesMentorAlone
  
    s1 = StudentAA.new(99)
    s2 = StudentAA.new(98)

    m = MentorAA.new("blah")
    m.set_students([s1,s2])
    studentP = ProgramAA.new
    mentorP = ProgramAA.new

    s2.set_program(studentP)
    m.set_program(mentorP)

    s1.delete

    assert_equal(1,m.number_of_students)
    assert_equal(mentorP,m.get_program)
    assert_equal(m,mentorP.get_mentor)
    assert_equal(studentP,s2.get_program)
    assert_equal(0,s1.number_of_mentors)
  end 

  def test_studentBounds
  
    assert_equal(2,MentorAA.maximum_number_of_students)
  end
end
end